### PR TITLE
positive-fix and pipeline-rework

### PR DESCRIPTION
**Pull request by Olenek**
Included a fix of collection of positive labeled dataset and rework of machine learning pipeline (introduction of ImbalancedLearn and sklearn2pmml for model inferrence)

### CODE REVIEW

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/PluginRunner.java [hunk 1]
```diff
@@ -62,7 +63,18 @@ private void runExtractions(CommandLine cmdLine) {
                 positiveRefactoringsExtractionRunner.run();
             }
         }
-
+        if (cmdLine.hasOption("generateNegativeSamples")) {
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/PluginRunner.java:66]: it is an exact copy of the code below, isn't it?
see lines 78-89

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/metrics/MetricCalculator.java [hunk 2]
```diff
@@ -30,28 +31,38 @@
 import static org.jetbrains.research.extractMethodExperiments.utils.StatementListUtil.ListToStr;
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/metrics/MetricCalculator.java:31]: unused import

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/metrics/MetricCalculator.java [hunk 3]
```diff
@@ -30,28 +31,38 @@
...
     private final PsiMethod method;
+    private final String repoPath;
+    private final String filePath;
     private final int beginLine;
     private final int endLine;
     private final FeaturesVector featuresVector;
 
-    public MetricCalculator(List<PsiStatement> statements, PsiMethod method, int beginLine, int endLine) {
-        this.statementsStr = ListToStr(statements);
+    public MetricCalculator(String code, PsiMethod method, int beginLine, int endLine) {
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/metrics/MetricCalculator.java:43]: do we need two constructors? we've already discussed it and agreed that there is no need for both of them. something changed?

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/haas/Candidate.java [hunk 4]
```diff
@@ -46,7 +49,7 @@ private void calculateRemainder() {
         ApplicationManager.getApplication().invokeAndWait(() -> {
             WriteCommandAction.runWriteCommandAction(originalMethod.getProject(), () -> {
                 for (PsiStatement statement : statementList) {
-                    statement.getParent().deleteChildRange(statementList.get(0), statementList.get(statementList.size() - 1));
+                    PsiTreeUtil.getParentOfType(statement, PsiCodeBlock.class).deleteChildRange(statementList.get(0), statementList.get(statementList.size() - 1));
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/haas/Candidate.java:52]: does it work as expected? I remember that we had problems with getting a parent for inner statements of` if `and `for` statements

Comment (onewhl) [reply RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/haas/Candidate.java:52]: btw, extract `PsiTreeUtil.getParentOfType(statement, PsiCodeBlock.class)` into a separate variable.
is it always an instance of `PsiCodeBlock`?

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/NegativeRefactoringsExtractionRunner.java [hunk 5]
```diff
@@ -112,16 +115,37 @@ private void writeFeaturesToFile(PsiFile psiFile, PsiMethod method, List<Candida
                 List<PsiStatement> statementList = candidate.getStatementList();
                 int beginLine = getNumberOfLine(psiFile, statementList.get(0).getTextRange().getStartOffset());
                 int endLine = getNumberOfLine(psiFile, statementList.get(statementList.size() - 1).getTextRange().getEndOffset());
-                MetricCalculator metricCalculator = new MetricCalculator(candidate.getStatementList(), method, beginLine, endLine);
+
+                Path tmpRepoPath = Paths.get(method.getContainingFile().getProject().getBasePath()).toAbsolutePath();
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/NegativeRefactoringsExtractionRunner.java:119]: why don't we use `psiFile.getProject().getName()` to get project name?

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/NegativeRefactoringsExtractionRunner.java [hunk 6]
```diff
@@ -112,16 +115,37 @@ private void writeFeaturesToFile(PsiFile psiFile, PsiMethod method, List<Candida
...
+        StringBuilder result = new StringBuilder();
+        for (PsiStatement statement : statementList) {
+            result.append(statement.getText());
+            result.append('\n');
+        }
+        return result.toString().strip();
+    }
 }
+
+
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/NegativeRefactoringsExtractionRunner.java:151]: extra line

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/NegativeRefactoringsExtractionRunner.java [hunk 7]
```diff
@@ -112,16 +115,37 @@ private void writeFeaturesToFile(PsiFile psiFile, PsiMethod method, List<Candida
...
                 this.fileWriter.append(String.valueOf(candidate.getScore()));
+                this.fileWriter.append(';');
+
+                this.fileWriter.append(repoName);
                 this.fileWriter.append('\n');
             }
         }
     }
+
+    public static String StatementListToStr(List<PsiStatement> statementList) {
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/NegativeRefactoringsExtractionRunner.java:141]: what about renaming with `statementsAsString`?

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 8]
```diff
@@ -1,26 +1,30 @@
 package org.jetbrains.research.extractMethodExperiments.extractors;
 
-import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.ide.highlighter.JavaFileType;
 import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vcs.VcsException;
 import com.intellij.openapi.vcs.changes.Change;
 import com.intellij.openapi.vfs.VirtualFile;
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:7]: unused import

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 9]
```diff
@@ -1,26 +1,30 @@
...
 
 import java.io.FileWriter;
 import java.io.IOException;
-import java.util.List;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
 import java.util.stream.Collectors;
 
 import static org.jetbrains.research.extractMethodExperiments.utils.PsiUtil.findMethodBySignature;
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:30]: unused import `import static org.jetbrains.research.extractMethodExperiments.utils.PsiUtil.getNumberOfLine;`

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 10]
```diff
@@ -66,45 +70,68 @@ private void handleCommit(List<Refactoring> refactorings) throws IOException {
...
+        if (changes.size() == 0) return;
+
+        for(Change change : changes){
+            try {
+                PsiFile sourcePsiFile = PsiFileFactory.getInstance(project).createFileFromText("tmp",
+                        JavaFileType.INSTANCE,
+                        change.getBeforeRevision().getContent());
+                changedSourceJavaFiles.put(change.getBeforeRevision().getFile().getPath(), sourcePsiFile);
+            } catch (VcsException e) {
+                e.printStackTrace();
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:88]: please, use another way to handle an exception.
log it

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 11]
```diff
@@ -66,45 +70,68 @@ private void handleCommit(List<Refactoring> refactorings) throws IOException {
...
-                .filter(f -> f.getVirtualFile() != null && f.getVirtualFile().getName().endsWith(".java"))
-                .map(Change::getVirtualFile)
-                .collect(Collectors.toList());
+        List<Change> changes = gitCommit.getChanges().stream().filter(f -> f.getVirtualFile() != null &&
+                f.getVirtualFile().getName().endsWith(".java")).collect(Collectors.toList());
+        Map<String, PsiFile> changedSourceJavaFiles = new HashMap<>();
+
+        if (changes.size() == 0) return;
+
+        for(Change change : changes){
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:81]: add a space after `for`

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 12]
```diff
@@ -66,45 +70,68 @@ private void handleCommit(List<Refactoring> refactorings) throws IOException {
...
 
+        this.fileWriter.append(repoName);
         this.fileWriter.append('\n');
     }
 
+    private static String getMethodSlice(PsiFile psiFile, int beginLine, int endLine) {
+        String[] fileLines = psiFile.getText().split("\n");
+        List<String> resultingLines = new ArrayList<>();
+        for (int i = 0; i < fileLines.length; i++){
+            if (i+1 >= beginLine && i+1 <= endLine) resultingLines.add(fileLines[i]);
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:132]: please, press `Ctrl + Alt + L` to format the code

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 13]
```diff
@@ -66,45 +70,68 @@ private void handleCommit(List<Refactoring> refactorings) throws IOException {
...
-            if (i != featuresVector.getDimension() - 1)
-                this.fileWriter.append(';');
+            this.fileWriter.append(';');
         }
 
+        this.fileWriter.append(repoName);
         this.fileWriter.append('\n');
     }
 
+    private static String getMethodSlice(PsiFile psiFile, int beginLine, int endLine) {
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:128]: why do we need this method? RefactoringMiner provides a method called `getExtractedCodeFragmentsFromSourceOperation()`.
why don't we use it?

**Review by onewhl — CHANGES_REQUESTED** [review]
Anton, thank you! Please, take a look at the comments I left.

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/PluginRunner.java (hunk 1 above)
Comment (Olenek) [reply RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/PluginRunner.java:66]: yeah, accidental misshap during tests of anticopypaster

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java (hunk 13 above)
Comment (Olenek) [reply RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:128]: it returns some hashset with very strange objects in it, while i need some simple text

**Review by Olenek — COMMENTED** [review]
At this stage, everything seems to work!

**Review by areyde — APPROVED** [review]
🔥

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 14]
```diff
@@ -66,45 +73,67 @@ private void handleCommit(List<Refactoring> refactorings) throws IOException {
...
+        String[] fileLines = psiFile.getText().split("\n");
+        List<String> resultingLines = new ArrayList<>();
+        for (int i = 0; i < fileLines.length; i++) {
+            if (i + 1 >= beginLine && i + 1 <= endLine) resultingLines.add(fileLines[i]);
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java:135]: please, move `resultingLines.add(fileLines[i]);` to a new line

File: RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java [hunk 15]
```diff
@@ -129,7 +129,8 @@ private static String getMethodSlice(PsiFile psiFile, int beginLine, int endLine
         String[] fileLines = psiFile.getText().split("\n");
         List<String> resultingLines = new ArrayList<>();
         for (int i = 0; i < fileLines.length; i++){
-            if (i+1 >= beginLine && i+1 <= endLine) resultingLines.add(fileLines[i]);
+            if (i+1 >= beginLine && i+1 <= endLine)
```
Comment (onewhl) [inline RefactoringsExperiments/src/main/java/org/jetbrains/research/extractMethodExperiments/extractors/CustomRefactoringHandler.java]: and add spaces 
i+1 -> i + 1

**Review by onewhl — APPROVED** [review]
Thank you!

**Pull request merged**